File: lib/agent/executor.ts
// Auto-executor: validates through risk gates, sizes via Kelly, executes LIVE

import type { AgentAnalysis } from './memory'
import { memory } from './memory'
import { evalAllGates, type GateResults } from './risk-gates'
import { readAgentConfig as readConfig } from './config-store'
import * as crypto from 'crypto'

export type ExecutionResult = {
  executed: boolean
  analysisId: string
  mode: string
  orderId?: string
  blockedBy?: string[]
  gateResults?: GateResults
  reason?: string
  sizeUSD?: number
  entryPx?: number
  stopPx?: number
  tpPx?: number
}

const HL_API = 'https://api.hyperliquid.xyz'

async function hlPost(body: object): Promise<unknown> {
  const res = await fetch(`${HL_API}/info`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) throw new Error(`HL API ${res.status}`)
  return res.json()
}

function kellySize(
  confidence: number,
  equity: number,
  rewardRiskRatio: number,
  maxTradeNotional: number,
): number {
  const p = confidence
  const q = 1 - p
  const b = rewardRiskRatio
  const fStar = Math.max(0, (p * b - q) / b)
  const halfKelly = fStar / 2
  const notional = halfKelly * equity
  return Math.min(notional, maxTradeNotional)
}

async function getMarketVolume24h(coin: string): Promise<number> {
  try {
    const majors = new Set(['BTC', 'ETH', 'SOL', 'BNB', 'XRP', 'DOGE', 'ADA', 'AVAX'])
    if (majors.has(coin)) return 100_000_000
    return 10_000_000
  } catch {
    return 0
  }
}

export async function maybeExecute(analysis: AgentAnalysis): Promise<ExecutionResult> {
  try {
    const config = await readConfig()
    const mode = (config.mode as string) || 'OFF'

    if (mode === 'OFF') {
      return { executed: false, mode, analysisId: analysis.id, reason: 'mode_off' }
    }

    let equity = 0
    let totalOpenNotional = 0
    let positions: Array<{ coin: string; side: string; sizeUSD: number }> = []
    const user = process.env.HYPERLIQUID_MASTER_ADDRESS || process.env.HYPERLIQUID_WALLET_ADDRESS || ''

    try {
      const [perpAcct, spotAcct] = await Promise.all([
        hlPost({ type: 'clearinghouseState', user }) as Promise<{
          marginSummary?: { accountValue: string; totalNtlPos: string }
          assetPositions?: Array<{ position: { coin: string; szi: string } }>
        }>,
        hlPost({ type: 'spotClearinghouseState', user }) as Promise<{
          balances?: Array<{ coin: string; total: string }>
        }>,
      ])

      const perpEquity = parseFloat(perpAcct.marginSummary?.accountValue ?? '0')
      totalOpenNotional = parseFloat(perpAcct.marginSummary?.totalNtlPos ?? '0')

      // Unified account: perp accountValue already includes spot collateral. 
      // Log spot balances for debugging.
      const spotBalances = (spotAcct.balances ?? [])
        .filter(b => ['USDC', 'USDT', 'USD'].includes(b.coin))
        .map(b => `${b.coin}: ${b.total}`)
        .join(', ') || 'none'
      console.log(`[executor] perp equity=$${perpEquity.toFixed(2)}, spot=${spotBalances}`)
      equity = perpEquity

      // Check perp positions for open position guard
      positions = (perpAcct.assetPositions ?? [])
        .filter(p => parseFloat(p.position.szi) !== 0)
        .map(p => ({
          coin: p.position.coin,
          side: parseFloat(p.position.szi) > 0 ? 'long' : 'short',
          sizeUSD: Math.abs(parseFloat(p.position.szi)) * (analysis.entryPx ?? 0),
        }))
    } catch (err) {
      console.error(`[executor] account fetch failed: ${err}`)
      return { executed: false, mode, analysisId: analysis.id, reason: 'account_fetch_failed' }
    }

    const dailyPnl = memory.getDailyPnl()

    const rewardRisk = analysis.tpPx && analysis.stopPx && analysis.entryPx
      ? Math.abs(analysis.tpPx - analysis.entryPx) / Math.abs(analysis.entryPx - analysis.stopPx)
      : 1.0
    const rawSize = kellySize(analysis.confidence, equity, rewardRisk, (config.maxTradeNotionalUsd as number) ?? 200)
    const tradeNotionalUSD = rawSize > 0 ? rawSize : (analysis.entryPx ?? 0) * 0.001

    const marketVolume24h = await getMarketVolume24h(analysis.coin)

    const recentTrades = memory.getRecentTrades(10)
    const lastTradeForCoin = recentTrades.find(t => t.coin === analysis.coin)
    const lastTradeTime = lastTradeForCoin?.executedAt

    const hasBinaryNews = analysis.newsContext
      ? /fed|fomc|cpi|rate|earnings|hack|exploit|SEC/i.test(analysis.newsContext)
      : false

    const ctx = {
      confidence: analysis.confidence,
      currentPositions: positions,
      tradeNotionalUSD: tradeNotionalUSD,
      dailyPnl,
      marketVolume24hUSD: marketVolume24h,
      coin: analysis.coin,
      tradeSide: (analysis.side ?? 'long') as 'long' | 'short',
      hasBinaryNewsRisk: hasBinaryNews,
      equity,
      totalOpenNotional,
    }

    const { results, blocked, blockReasons } = evalAllGates(ctx, config, lastTradeTime)

    if (blocked) {
      memory.recordTrade({
        id: crypto.randomUUID(),
        analysisId: analysis.id,
        coin: analysis.coin,
        side: (analysis.side ?? 'long') as 'long' | 'short',
        entryPx: analysis.entryPx ?? 0,
        sizeUSD: 0,
        executedAt: Date.now(),
      })

      return {
        executed: false,
        mode,
        analysisId: analysis.id,
        blockedBy: blockReasons,
        gateResults: results,
      }
    }

    // LIVE mode — real orders only
    const walletKey = process.env.HYPERLIQUID_PRIVATE_KEY
    if (!walletKey) {
      return { executed: false, mode, analysisId: analysis.id, reason: 'private_key_missing' }
    }

    const orderRes = await fetch(
      `${process.env.NEXT_PUBLIC_BASE_URL || 'http://localhost:3000'}/api/hl/place-order`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          side: analysis.side === 'long' ? 'long' : 'short',
          riskUSD: tradeNotionalUSD,
          leverage: 5,
          coin: analysis.coin,
        }),
      },
    )

    const orderData = await orderRes.json() as { ok: boolean; orderId?: string; error?: string }

    if (!orderData.ok) {
      return {
        executed: false,
        mode,
        analysisId: analysis.id,
        reason: `order_failed: ${orderData.error ?? 'unknown'}`,
        gateResults: results,
      }
    }

    memory.recordTrade({
      id: crypto.randomUUID(),
      analysisId: analysis.id,
      coin: analysis.coin,
      side: (analysis.side ?? 'long') as 'long' | 'short',
      entryPx: analysis.entryPx ?? 0,
      sizeUSD: tradeNotionalUSD,
      orderId: orderData.orderId,
      executedAt: Date.now(),
    })

    return {
      executed: true,
      mode,
      analysisId: analysis.id,
      orderId: orderData.orderId,
      gateResults: results,
      sizeUSD: tradeNotionalUSD,
      entryPx: analysis.entryPx ?? 0,
      stopPx: analysis.stopPx,
      tpPx: analysis.tpPx,
    }
  } catch (err) {
    return {
      executed: false,
      mode: 'ERROR',
      analysisId: analysis.id,
      reason: err instanceof Error ? err.message : String(err),
    }
  }
}

File: hermes_agent/agents/executor.py
"""Auto-executor: validates through risk gates, sizes the trade, executes LIVE.

Integrates the DSL exit engine for two-phase trailing stops
(loss protection -> profit locking).
"""

from __future__ import annotations

import logging
import os
import re
import time
import uuid
from typing import Any, Dict, List

from hermes_agent.agents.config_store import read_agent_config
from hermes_agent.agents.dsl_exit import ExitPolicy, check_all_positions, register_position
from hermes_agent.agents.memory import memory
from hermes_agent.agents.risk_gates import GateContext, eval_all_gates
from hermes_agent.client.exchange import (
    HL_LEVERAGE,
    get_hl_atr,
    get_hl_price,
    place_hl_order,
    place_hl_trigger_order,
    set_leverage,
)
from hermes_agent.client.hl_client import fetch_account_state, resolve_user_address

logger = logging.getLogger(__name__)

SL_ATR_MULT = 3.5
TP_ATR_MULT = 1.0

# Default 24h volumes for coins not in the major list
_MAJOR_VOLUMES = {
    "BTC": 1e8, "ETH": 1e8, "SOL": 1e8, "BNB": 1e8,
    "XRP": 1e8, "DOGE": 1e8, "ADA": 1e8, "AVAX": 1e8,
}


def _get_market_volume_24h(coin: str) -> float:
    return _MAJOR_VOLUMES.get(coin, 1e7)


def kelly_size(
    confidence: float,
    equity: float,
    reward_risk_ratio: float,
    max_trade_notional: float,
) -> float:
    """Calculate trade size using the half-Kelly criterion."""
    p = confidence
    q = 1 - p
    b = reward_risk_ratio
    f_star = max(0, (p * b - q) / b) if b != 0 else 0
    half_kelly = f_star / 2
    notional = half_kelly * equity
    return min(notional, max_trade_notional)


def maybe_execute(analysis: Dict[str, Any]) -> Dict[str, Any]:
    """Execute an analysis through risk gates and into the market."""
    config = read_agent_config()
    mode = str(config.get("mode", "OFF"))

    if mode == "OFF":
        return {
            "executed": False, "mode": mode,
            "analysis_id": analysis["id"], "reason": "mode_off",
        }

    # Idempotency: don't double-execute
    already = next(
        (t for t in memory.get_recent_trades(100)
         if t.get("analysis_id") == analysis["id"] and t.get("size_usd", 0) > 0),
        None,
    )
    if already:
        return {
            "executed": False, "mode": mode,
            "analysis_id": analysis["id"], "reason": "already_executed",
            "order_id": already.get("order_id"),
        }

    user = resolve_user_address()
    state = fetch_account_state(user)
    equity = state["equity"]
    total_open_notional = state["total_ntl"]

    memory.track_daily_pnl(equity)
    daily_pnl = memory.get_daily_pnl()

    positions = [
        {
            "coin": p["position"]["coin"],
            "side": "long" if float(p["position"]["szi"]) > 0 else "short",
            "size_usd": abs(float(p["position"]["szi"])) * (analysis.get("entry_px") or 0),
        }
        for p in state["asset_positions"]
    ]

    entry_px = analysis.get("entry_px")
    tp_px = analysis.get("tp_px")
    stop_px = analysis.get("stop_px")

    # Notional sized off equity (1% * leverage), overriding any AI-provided value.
    trade_notional = equity * 0.01 * HL_LEVERAGE

    recent_trades = memory.get_recent_trades(10)
    last_trade = next(
        (t for t in recent_trades if t.get("coin") == analysis["coin"]),
        None,
    )
    last_trade_time = last_trade.get("executed_at") if last_trade else None

    has_binary_news = bool(
        analysis.get("news_context")
        and re.search(
            r"fed|fomc|cpi|rate|earnings|hack|exploit|SEC",
            analysis["news_context"],
            re.IGNORECASE,
        )
    )

    trade_side = analysis.get("side", "long") or "long"
    ctx = GateContext(
        confidence=analysis["confidence"],
        current_positions=positions,
        trade_notional_usd=trade_notional,
        daily_pnl=daily_pnl,
        market_volume_24h_usd=_get_market_volume_24h(analysis["coin"]),
        coin=analysis["coin"],
        trade_side=trade_side,
        has_binary_news_risk=has_binary_news,
        equity=equity,
        total_open_notional=total_open_notional,
    )

    gate_output = eval_all_gates(ctx, config, last_trade_time)

    if gate_output["blocked"]:
        memory.record_trade({
            "id": str(uuid.uuid4()),
            "analysis_id": analysis["id"],
            "coin": analysis["coin"],
            "side": trade_side,
            "entry_px": entry_px or 0,
            "size_usd": 0,
            "executed_at": int(time.time() * 1000),
        })
        return {
            "executed": False, "mode": mode,
            "analysis_id": analysis["id"],
            "blocked_by": gate_output["block_reasons"],
            "gate_results": gate_output["results"],
        }

    if not os.environ.get("HYPERLIQUID_PRIVATE_KEY"):
        return {
            "executed": False, "mode": mode,
            "analysis_id": analysis["id"],
            "reason": "private_key_missing",
        }

    coin = analysis["coin"]
    is_buy = trade_side == "long"

    # Fetch live mid — never use the (possibly stale) analysis entry price.
    mid_price = get_hl_price(coin)
    if mid_price <= 0:
        return {"executed": False, "mode": mode, "analysis_id": analysis["id"],
                "reason": f"invalid_price_for_{coin}"}

    # Size in coin from the leverage-inclusive notional, floored at HL's $10
    # minimum order value and capped at 100 coins to bound very cheap assets.
    size_in_coin = trade_notional / mid_price
    min_size_by_value = 10.0 / mid_price
    size_in_coin = max(size_in_coin, min_size_by_value)
    size_in_coin = min(size_in_coin, 100.0)

    position_notional = trade_notional

    atr = get_hl_atr("4h", 14, coin)

    set_leverage(coin, HL_LEVERAGE)
    order_res = place_hl_order(is_buy, size_in_coin, mid_price, coin)

    if not order_res.get("ok"):
        return {
            "executed": False, "mode": mode, "analysis_id": analysis["id"],
            "reason": f"order_failed: {order_res.get('error', 'unknown')}",
            "gate_results": gate_output["results"],
        }

    # Register the position with the DSL tracker; it re-evaluates the exit
    # floor on every scan tick (loss protection -> profit locking).
    dsl_config = config.get("dsl_exit", {})
    policy = ExitPolicy(
        max_loss_pct=dsl_config.get("max_loss_pct", 2.5),
        protect_pct=dsl_config.get("protect_pct", 1.5),
        retrace_threshold=dsl_config.get("retrace_threshold", 0.30),
        hard_timeout_minutes=dsl_config.get("hard_timeout_minutes", 180.0),
    )
    register_position(coin, trade_side, mid_price, policy=policy)
    logger.info(f"[executor] Registered DSL exit for {coin} {trade_side} @ {mid_price}")

    memory.record_trade({
        "id": str(uuid.uuid4()),
        "analysis_id": analysis["id"],
        "coin": coin,
        "side": trade_side,
        "entry_px": mid_price,
        "size_usd": position_notional,
        "order_id": order_res.get("order_id"),
        "executed_at": int(time.time() * 1000),
    })

    # Backup exchange stop-loss bracket — DSL is the primary exit engine.
    if atr > 0 and size_in_coin > 0:
        sl_px = mid_price - atr * SL_ATR_MULT if is_buy else mid_price + atr * SL_ATR_MULT
        sl_res = place_hl_trigger_order(is_buy, size_in_coin, sl_px, "sl", coin)
        if sl_res.get("ok"):
            logger.info(f"[executor] Placed backup SL at {sl_px}")
        else:
            logger.error(f"[executor] Backup SL FAILED for {coin}: {sl_res.get('error')}")

    final_sl = (mid_price - atr * SL_ATR_MULT) if is_buy else (mid_price + atr * SL_ATR_MULT) if atr > 0 else stop_px
    final_tp = (mid_price + atr * TP_ATR_MULT) if is_buy else (mid_price - atr * TP_ATR_MULT) if atr > 0 else tp_px

    return {
        "executed": True, "mode": mode,
        "analysis_id": analysis["id"],
        "order_id": order_res.get("order_id"),
        "gate_results": gate_output["results"],
        "size_usd": position_notional,
        "entry_px": mid_price,
        "stop_px": final_sl,
        "tp_px": final_tp,
        "dsl_registered": True,
    }


def monitor_exits(mids: Dict[str, float]) -> List[Dict[str, Any]]:
    """Check all DSL-tracked positions and return those that should be closed."""
    exits = check_all_positions(mids)
    return [
        {
            "coin": v.coin,
            "side": v.phase,
            "reason": v.reason,
            "unrealized_pct": v.unrealized_pct,
        }
        for v in exits
    ]
